### ISSUE TITLE
States rebuilt when index changed and state is different in ListView.builder (ListView.custom) 

### ISSUE THREAD

**Issue by Tom3652** [flutter/flutter#100451]
## Steps to Reproduce

1. Run the sample code
2. Click on the floating button as many times as you want and watch the logs
3. See that all items are disposed (in the logs) and recreated after each insert

**Expected results:**

The widget state should be kept even when changing index (for different states). 

**Actual results:** 

The widgets should not be disposed after inserting new items and changing the "states rank".  

This is very hard to implement a usual chat app with this issue as we have many `StatefulWidgets` of different kind and we must have a `reversed` and ordered list as you probably see in your daily apps.  

This issue reloads all of the widgets where the state is changing for a given index.  

If I am doing a wrong implementation please feel free to let me know.  
I have also checked `ListView.custom` and observed the same behavior (despite it was suggested in other similar issues). 

<details>
<summary>Code sample</summary>

### `ListView.builder` : 

```dart

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: ListDisposeTest()
      ),
    );
  }
}

class ListDisposeTest extends StatefulWidget {
  const ListDisposeTest({Key? key}) : super(key: key);

  @override
  State<ListDisposeTest> createState() => _ListDisposeTestState();
}

class _ListDisposeTestState extends State<ListDisposeTest> {
  List<int> numberList = [];

  int number = -1;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      floatingActionButton: FloatingActionButton(
       child: Icon(
         Icons.add
        ),
        onPressed: () {
          setState(() {
            number += 1;
            numberList.insert(0, number);
          });
        },
      ),
      body: Stack(
        children: [
          SafeArea(
            child: ListView.builder(
                itemBuilder: (context, index) {
                  return numberList[index] % 2 == 0
                      ? ItemListDisposePair(
                          index: numberList[index],
                          key: ValueKey<int>(numberList[index]))
                      : ItemListDispose(
                          index: numberList[index],
                          key: ValueKey<int>(numberList[index]));
                },
                reverse: true,
                itemCount: numberList.length),
          ),
        ],
      ),
    );
  }
}

class ItemListDispose extends StatefulWidget {
  final int index;

  const ItemListDispose({Key? key, required this.index}) : super(key: key);

  @override
  State<ItemListDispose> createState() => _ItemListDisposeState();
}

class _ItemListDisposeState extends State<ItemListDispose> {
  @override
  void initState() {
    print("init item : ${widget.index}");
    super.initState();
  }

  @override
  void dispose() {
    print("dispose item : ${widget.index}");
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.red,
      margin: EdgeInsets.only(bottom: 20),
      height: 50,
      child: Text(
        "${widget.index} + impair",
        style: TextStyle(color: Colors.white, fontSize: 20),
      ),
    );
  }
}

class ItemListDisposePair extends StatefulWidget {
  final int index;

  const ItemListDisposePair({Key? key, required this.index}) : super(key: key);

  @override
  State<ItemListDisposePair> createState() => _ItemListDisposePairState();
}

class _ItemListDisposePairState extends State<ItemListDisposePair> {
  @override
  void initState() {
    print("init item : ${widget.index}");
    super.initState();
  }

  @override
  void dispose() {
    print("dispose item : ${widget.index}");
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.blue,
      margin: EdgeInsets.only(bottom: 20),
      height: 50,
      child: Text(
        "${widget.index} + pair ",
        style: TextStyle(color: Colors.white, fontSize: 20),
      ),
    );
  }
}
```

</details>

<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel stable, 2.10.3, on macOS 12.3 21E230 darwin-x64, locale fr-FR)
    • Flutter version 2.10.3 at /Users/foxtom/Desktop/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7e9793dee1 (3 weeks ago), 2022-03-02 11:23:12 -0600
    • Engine revision bd539267b4
    • Dart version 2.16.1
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/foxtom/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)

[✓] VS Code (version 1.65.2)
    • VS Code at /Users/foxtom/Desktop/Visual Studio Code.app/Contents
    • Flutter extension version 3.36.0

[✓] Connected device (3 available)
    • moto g 8 power (mobile) • ZY22BNDW2C • android-arm64  • Android 11 (API 30)
    • macOS (desktop)         • macos      • darwin-x64     • macOS 12.3 21E230 darwin-x64
    • Chrome (web)            • chrome     • web-javascript • Google Chrome 99.0.4844.83

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by darshankawar** [comment]
@Tom3652 
Since you are calling set state to change the number, it causes the widget to rebuild:
```

onPressed: () {
          setState(() {
            number += 1;
            numberList.insert(0, number);
          });
        },
```

So I guess this is an expected behavior.

**Comment by Tom3652** [comment]
If this is the natural behavior, then how can we keep the state of widgets ? 

Because this would mean `AutomaticKeepAliveMixin` has not the desired behavior, returning true and disposing the `Widget` anyway :(

(It actually works by not disposing widgets that are added at a higher rank.)

The reload pattern is :

- Widget `X` with `State A` at rank `n`
- Widget `Y` with `State B` at rank `n+1`
- Insert a widget `Z` with `State A` at rank `n`
-> Widget `X` goes at rank `n+1`
-> Old widget (`Y`) at rank `n+1` is a `State B` widget, so the widget at rank `n+1` is destroyed and recreated to display a `State A` instead of `State B`.

So the question is, is there any way to make the List reorder the item state on insert at an existing rank ? 

I understand the implementation made with `StatefulWidgets` in `ListView.builder` but this make a chat app very hard to build when different kind of widget `States` are disposed and recreated all the tie, including audio, videos, text etc...

Adding `ValueKey`s don't seem to help solving this either, so i am not sure if this is a bug or if this the intented behavior, but if it is it's really problematic :( 

Also, in my example i am using `SetState` but in my real app i am using `providers` but this doesn't have anything to do with the internal behavior of the `ListView` in my humble opinion. 

Can you confirm please that this is the intended behavior and there is no workaround ?

Thanks a lot

**Comment by darshankawar** [comment]
> is there any way to make the List reorder the item state on insert at an existing rank ?

Do you mean to maintain a scroll position on insert in a listview ?

**Comment by Tom3652** [comment]
Well technically it's not really maintaining a scroll position in a way that i don't care if there is an actual scroll, it's especially about widget states that are staying at the same index no matter what.   

So if there is a change for a specific index (if you insert in the middle of the list for example), and the new widget has a different state than the one at this same index before, then the ListView will destroy the previous ones instead of "putting the existing state at the new index".   
This means rebuilding all of the items to "move them" after / before the new inserted item. But the "move" is done by destroying and rebuilding instead of an actual "moving".

Please tell me if this is not clear, i will try to make screenshots / video records if needed.

**Comment by darshankawar** [comment]
Thanks for the feedback. Although I do see the same behavior as reported, I am not sure if this is intended behavior or a bug in the framework. I will label it for further insights from the team.

/cc @Piinks 

<details>
<summary> stable, master flutter doctor -v </summary>

```

[✓] Flutter (Channel stable, 2.10.3, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 2.10.3 at /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7e9793dee1 (11 hours ago), 2022-03-02 11:23:12 -0600
    • Engine revision bd539267b4
    • Dart version 2.16.1
    • DevTools version 2.9.2

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.


[✓] Flutter (Channel master, 2.13.0-0.0.pre.191, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 2.13.0-0.0.pre.191 at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 69b92f0452 (3 minutes ago), 2022-03-21 21:25:09 -0700
    • Engine revision e6233bd30b
    • Dart version 2.17.0 (build 2.17.0-223.0.dev)
    • DevTools version 2.11.4

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.


```

</details>

**Comment by xu-baolin** [comment]
@Tom3652, hi, thanks for reporting.
This should be fixed after #100547 landed.

**Comment by Tom3652** [comment]
Thank you very much for the so fast fix, much appreciated !

**Comment by xu-baolin** [comment]
> Thank you very much for the so fast fix, much appreciated !

`ListView.custom` already works if you provider the `SliverChildDelegate.findIndexByKey`.

**Comment by Tom3652** [comment]
Hi ! 
Was this merged and release in stable `flutter: 2.10.4` ? Or only in `master` channel not yet stable ?
